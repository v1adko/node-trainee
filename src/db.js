const { flags, connectionDBString } = require('./config');
const mongoose = require('mongoose');

mongoose.Promise = global.Promise;
if (flags.debug) {
  mongoose.set('debug', true);
}

let db = null;

const connect = () => {
  if (db === null) {
    db = mongoose.connect(connectionDBString, {
      useMongoClient: true
    });
  } else {
    throw new Error('DB instance already exists, use existing connection or close it before creating a new one.');
  }
};

const getConnection = () => {
  if (db) {
    return db;
  }
  throw new Error("DB connection doesn't exist yet.");
};

const close = () => {
  if (db) {
    db.close(() => {
      console.log('Mongoose disconnected on app');
    });
  } else {
    throw new Error("DB connection doesn't exist yet.");
  }
};

module.exports = {
  connect,
  getConnection,
  close
};

