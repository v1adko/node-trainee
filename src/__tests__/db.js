import db from '../db';

describe('Test db set connection', () => {
  it('should create db connection', () => {
    const connection = db.connect();

    expect(connection).not.toBeUndefined();

    db.closeConnection();
  });

  it('should not create db connection because it already exist, and should return exist db connection', () => {
    const connection = db.connect();

    const secondConnection = db.connect();

    expect(connection).toEqual(secondConnection);

    db.closeConnection();
  });

  it('should close db connection, and then throw error in trying to get connection', () => {
    db.connect();
    db.closeConnection();

    expect(db.getConnection).toThrowErrorMatchingSnapshot();
  });

  it("should not close db connection because it doesn't exist", () => {
    expect(db.closeConnection).toThrowErrorMatchingSnapshot();
  });

  it('should get db connection', () => {
    const cretedConnection = db.connect();

    const gettedConnection = db.getConnection();

    expect(cretedConnection).toEqual(gettedConnection);

    db.closeConnection();
  });

  it('should reopen db connection', () => {
    const connection = db.connect();

    const reopenedConnection = db.tryReopen();

    expect(connection).toEqual(reopenedConnection);

    db.closeConnection();
  });
});
