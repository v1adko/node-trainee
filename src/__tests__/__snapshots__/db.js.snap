// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test db set connection should close db connection, and then throw error in trying to get connection 1`] = `"DB connection doesn't exist yet."`;

exports[`Test db set connection should not close db connection because it doesn't exist 1`] = `"DB connection doesn't exist yet."`;
